packages/style-engine/class-wp-style-engine.php

```diff
@@ -370,6 +386,12 @@ protected static function get_css_declarations( $style_value, $style_definition,
 					$value = static::get_css_var_value( $value, $style_definition['css_vars'] );
 				}
 				$individual_property = sprintf( $style_property_keys['individual'], _wp_to_kebab_case( $key ) );
+
+				// If the style value contains a reference to another value in the tree.
+				if ( isset( $value['ref'] ) ) {
+					$value = static::get_ref_value( $block_styles, $value['ref'] );
```

Comment:
Does this assume that `$block_styles` will always contain the whole tree? I was wondering if for `get_ref_value` we might need an additional param for the complete tree (for when we add in elements / block level styles)?

Reply:
Yes, it'd have to in this scenario. 

It's another argument for the one-node-at-a-time approach [discussed above](https://github.com/WordPress/gutenberg/pull/42143#discussion_r918562149): letting WP_Theme_JSON do the ref value fetching for us.

Reply:
If we want to have a method that accepts an entire theme.json styles tree, then it could be a separate one where we'd assume that the entire tree is passed, e.g., `wp_style_engine_generate_stylesheet` or 🤷‍♂️

Reply:
> It's another argument for the one-node-at-a-time approach https://github.com/WordPress/gutenberg/pull/42143#discussion_r918562149: letting WP_Theme_JSON do the ref value fetching for us.

It's so fascinating these sorts of problems: even just the wording of your comment there makes it feel clearer that the theme JSON class is probably a better place to do ref lookups (as the owner of that data structure), and the style engine should probably treat whatever it's given as the "real" thing. I don't think there's any real right or wrong way to go about it, but these sorts of subtle distinctions really do help us tease apart what the style engine is and isn't! (Or how much scope we're comfortable taking on at any one time 😄)